plugins/s3_auth/aws_auth_v4.cc

```diff
@@ -0,0 +1,730 @@
+#include <iostream>
+#endif
+
+#include "aws_auth_v4.h"
+
+/**
+ * @brief Lower-case Base16 encode a character string (hexadecimal format)
+ *
+ * @see AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+ * Base16 RFC4648: https://tools.ietf.org/html/rfc4648#section-8
+ *
+ * @param in ptr to an input counted string to be base16 encoded.
+ * @param inLen input character string length
+ * @return base16 encoded string.
+ */
+String
+base16Encode(const char *in, size_t inLen)
+{
+  if (nullptr == in || inLen == 0) {
+    return String();
+  }
+
+  std::stringstream result;
+
+  const char *src    = in;
+  const char *srcEnd = in + inLen;
+
+  while (src < srcEnd) {
+    result << std::setfill('0') << std::setw(2) << std::hex << static_cast<int>((*src) & 0xFF);
+    src++;
+  }
+  return result.str();
+}
+
+/**
+ * @brief URI-encode a character string (AWS specific version, see spec)
+ *
+ * @see AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+ *
+ * @param in string to be URI encoded
+ * @param isObjectName if true don't encode '/', keep it as it is.
+ * @return encoded string.
+ */
+String
+uriEncode(const String in, bool isObjectName)
+{
+  std::stringstream result;
+
+  for (std::string::size_type i = 0; i < in.length(); i++) {
+    if (isalnum(in[i]) || in[i] == '-' || in[i] == '_' || in[i] == '.' || in[i] == '~') {
+      /* URI encode every byte except the unreserved characters:
+       * 'A'-'Z', 'a'-'z', '0'-'9', '-', '.', '_', and '~'. */
+      result << in[i];
+    } else if (in[i] == ' ') {
+      /* The space character is a reserved character and must be encoded as "%20" (and not as "+"). */
+      result << "%20";
+    } else if (isObjectName && in[i] == '/') {
+      /* Encode the forward slash character, '/', everywhere except in the object key name. */
+      result << "/";
+    } else {
+      /* Letters in the hexadecimal value must be upper-case, for example "%1A". */
+      result << "%" << std::uppercase << std::setfill('0') << std::setw(2) << std::hex << (int)in[i];
+    }
+  }
+
+  return result.str();
+}
+
+/**
+ * @brief lower case a C string
+ *
+ * @param in ptr to the char string to be converted
+ * @param inLen length of the string to be converted
+ * @return a string converted to lower case.
+ */
+String
+lowercase(const char *in, size_t inLen)
+{
+  if (nullptr == in || inLen == 0) {
+    return String();
+  }
+
+  std::stringstream result;
+
+  std::locale loc;
+  for (std::string::size_type i = 0; i < inLen; ++i) {
+    result << std::tolower(*(in + i), loc);
+  }
+
+  return result.str();
+}
+
+/**
+ * @brief lower case a C++ string
+ *
+ * @param in ptr to the char string to be converted
+ * @param inLen length of the string to be converted
+ * @return a string converted to lower case.
+ */
+String
+lowercase(const String &in)
+{
+  std::stringstream result;
+
+  std::locale loc;
+  for (auto i : in) {
```

Comment:
@zwoop, @SolidWallOfCode: realized that both `lowercase()` functions are unnecessary (implemented them first since I thought I would use them a lot, but now after I finished realizing that they would not lead to better code reuse). Removed both, in-place `transform()` fits better in both places, thank you!